ksqldb-engine/src/main/java/io/confluent/ksql/engine/KsqlEngine.java

```diff
@@ -293,6 +316,186 @@ public TransientQueryMetadata executeTransientQuery(
+      } else {
+        try {
+          Thread.sleep(50L);
+        } catch (final InterruptedException e) {
+          log.error("Interrupted waiting for stream pull query to complete", e);
+          throw new KsqlServerException("Interrupted");
+        }
+      }
+    }
+    transientQueryMetadata.getKafkaStreams().close(Duration.ZERO);
+    QueryLogger.info("Finished pull query results", statement.getStatementText());
+  }
+
+  private TopicDescription getTopicDescription(final Admin admin, final String sourceTopicName) {
+    final KafkaFuture<TopicDescription> topicDescriptionKafkaFuture = admin
+        .describeTopics(Collections.singletonList(sourceTopicName))
+        .values()
+        .get(sourceTopicName);
+
+    try {
+      return topicDescriptionKafkaFuture.get(10, TimeUnit.SECONDS);
+    } catch (final InterruptedException e) {
+      log.error("Admin#describeTopics(" + sourceTopicName + ") interrupted", e);
+      throw new KsqlServerException("Interrupted");
+    } catch (final ExecutionException e) {
+      log.error("Error executing Admin#describeTopics(" + sourceTopicName + ")", e);
+      throw new KsqlServerException("Internal Server Error");
+    } catch (final TimeoutException e) {
+      log.error("Admin#describeTopics(" + sourceTopicName + ") timed out", e);
+      throw new KsqlServerException("Backend timed out");
+    }
+  }
+
+  private Map<TopicPartition, Long> getEndOffsets(
+      final Admin admin,
+      final TopicDescription topicDescription,
+      final IsolationLevel isolationLevel) {
+    final Map<TopicPartition, OffsetSpec> topicPartitions =
+        topicDescription
+            .partitions()
+            .stream()
+            .map(td -> new TopicPartition(topicDescription.name(), td.partition()))
+            .collect(toMap(identity(), tp -> OffsetSpec.latest()));
+
+    final ListOffsetsResult listOffsetsResult = admin.listOffsets(
+        topicPartitions,
+        new ListOffsetsOptions(
+            isolationLevel
+        )
+    );
+
+    try {
+      final Map<TopicPartition, ListOffsetsResult.ListOffsetsResultInfo> partitionResultMap =
+          listOffsetsResult.all().get(10, TimeUnit.SECONDS);
+      return partitionResultMap
+          .entrySet()
+          .stream()
+          .collect(toMap(Map.Entry::getKey, e -> e.getValue().offset()));
+    } catch (final InterruptedException e) {
+      log.error("Admin#listOffsets(" + topicDescription.name() + ") interrupted", e);
+      throw new KsqlServerException("Interrupted");
+    } catch (final ExecutionException e) {
+      log.error("Error executing Admin#listOffsets(" + topicDescription.name() + ")", e);
+      throw new KsqlServerException("Internal Server Error");
+    } catch (final TimeoutException e) {
+      log.error("Admin#listOffsets(" + topicDescription.name() + ") timed out", e);
+      throw new KsqlServerException("Backend timed out");
+    }
+  }
+
+  private boolean passedEndOffsets(final Admin admin,
```

Comment:
Eh, I tried this refactor out, and I think it makes the code slightly more complicated to follow, since we then have to book-keep which partitions we're done with. Since the perf improvement is marginal, I'll just keep this the same.

Reply:
👍